<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>4.5. Smoothing raw tag count data across the genome &mdash; BEDOPS 2.4.1 documentation</title>
    
    <link rel="stylesheet" href="../../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../',
        VERSION:     '2.4.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../../_static/sidebar.js"></script>
    <link rel="shortcut icon" href="../../_static/favicon.ico"/>
    <link rel="top" title="BEDOPS 2.4.1 documentation" href="../../index.html" />
    <link rel="up" title="4. Usage examples" href="../usage-examples.html" />
    <link rel="next" title="4.6. Efficiently creating Starch-formatted archives with a cluster" href="starchcluster.html" />
    <link rel="prev" title="4.4. Finding the subset of SNPs within DHSes" href="snps-within-dhses.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="starchcluster.html" title="4.6. Efficiently creating Starch-formatted archives with a cluster"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="snps-within-dhses.html" title="4.4. Finding the subset of SNPs within DHSes"
             accesskey="P">previous</a> |</li>
        <li><a href="../../index.html">BEDOPS 2.4.1 documentation</a> &raquo;</li>
          <li><a href="../usage-examples.html" accesskey="U">4. Usage examples</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="smoothing-raw-tag-count-data-across-the-genome">
<span id="smoothing-raw-tags"></span><h1>4.5. Smoothing raw tag count data across the genome<a class="headerlink" href="#smoothing-raw-tag-count-data-across-the-genome" title="Permalink to this headline">¶</a></h1>
<p>In this example, we generate smoothed density signal by binning the genome into 20 bp intervals and counting the number of non-paired-end tag reads falling within 75 bp of each interval. A simple follow-on script marks up results to wig or bigWig format for loading into a track of a local UCSC Genome Browser.</p>
<div class="section" id="bedops-tools-in-use">
<h2>4.5.1. BEDOPS tools in use<a class="headerlink" href="#bedops-tools-in-use" title="Permalink to this headline">¶</a></h2>
<p>For this script, we use <a class="reference internal" href="../reference/file-management/conversion/bam2bed.html#bam2bed"><em>bam2bed</em></a> to convert a BAM file to BED, then we use <a class="reference internal" href="../reference/statistics/bedmap.html#bedmap"><em>bedmap</em></a> to run a sliding density window over input genomic regions. Finally <a class="reference internal" href="../reference/file-management/compression/starch.html#starch"><em>starch</em></a> compresses the results.</p>
</div>
<div class="section" id="script">
<h2>4.5.2. Script<a class="headerlink" href="#script" title="Permalink to this headline">¶</a></h2>
<div class="highlight-python"><pre>#!/bin/tcsh -ef
# author : Richard Sandstrom

if ( $#argv != 5 ) then
  printf "Wrong number of arguments\n"
  printf "&lt;bam-file&gt; &lt;out-file&gt; &lt;window-size&gt; &lt;step-size&gt; &lt;chromosome-file&gt;\n"
  printf "  where &lt;chromosome-file&gt; contains whole chromosome BED items for the\n"
  printf "  genome, e.g., sort-bed formatted output from the UCSC hg19.chromInfo table.\n"
  exit -1
endif

# BAM file
set inBam = $argv[1]
# resulting density file
set outDensity = $argv[2]
# +/- window for counting read 5' ends
set window = $argv[3]
# step size across genome
set binI = $argv[4]
# chromosome file for organism of interest
set chromsfile = $argv[5]

set outDir = $outDensity:h
mkdir -p $outDir

set tmpDir = /tmp/`whoami`/scratch/$$
if ( -d $tmpDir ) then
  rm -rf $tmdDir
endif
mkdir -p $tmpDir

# clip tags to single 5' end base
bam2bed &lt; $inBam \
    | awk '{if($6=="+"){s=$2; e=$2+1}else{s=$3-1; e=$3}print $1"\t"s"\t"e}' \
    | sort-bed - \
   &gt;! $tmpDir/tags.bed

# create genome-wide bins and count how many tags fall within range of each
awk -v binI=$binI -v win=$window \
      '{ \
         i = 0; \
         for(i = $2; i &lt;= $3-binI; i += binI) { print $1"\t"i"\t"i + binI } \
         # end of chrome may include a bin of size &lt; binI \
         if ( i &lt; $3 ) { print $1"\t"i"\t"$3; } \
      }' $chromsfile \
    | bedmap --faster --range $window --echo --count --delim "\t" - $tmpDir/tags.bed \
    | starch - \
   &gt;! $outDensity

rm -rf $tmpDir

exit 0</pre>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../../index.html">
              <img class="logo" src="../../_static/logo_with_label_v2.png" alt="Logo"/>
            </a></p>
  <h3><a href="../../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">4.5. Smoothing raw tag count data across the genome</a><ul>
<li><a class="reference internal" href="#bedops-tools-in-use">4.5.1. BEDOPS tools in use</a></li>
<li><a class="reference internal" href="#script">4.5.2. Script</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="snps-within-dhses.html"
                        title="previous chapter">4.4. Finding the subset of SNPs within DHSes</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="starchcluster.html"
                        title="next chapter">4.6. Efficiently creating Starch-formatted archives with a cluster</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="../../_sources/content/usage-examples/smoothing-tags.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="starchcluster.html" title="4.6. Efficiently creating Starch-formatted archives with a cluster"
             >next</a> |</li>
        <li class="right" >
          <a href="snps-within-dhses.html" title="4.4. Finding the subset of SNPs within DHSes"
             >previous</a> |</li>
        <li><a href="../../index.html">BEDOPS 2.4.1 documentation</a> &raquo;</li>
          <li><a href="../usage-examples.html" >4. Usage examples</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2014, Shane Neph, Alex Reynolds.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.
    </div>
  </body>
</html>